syntax = "proto3";

package officepb;
option go_package = "proto";

message User {
  string name = 1;
  string password = 2;
  string bio = 3;
}

message Document {
  int64 id = 1;
  string token = 2;
  string name = 3;
  Ctx context = 4;
  string content = 5;
}

message Ctx {
  map<string, string> vars = 1;
  repeated Expr exprs = 2;
}

message Expr {
  string name = 1;
  string expr = 2;
}

message RegisterRequest {
  string name = 1;
  string password = 2;
  string bio = 3;
}

message RegisterResponse {
  string session = 1;
}

message LoginRequest {
  string name = 1;
  string password = 2;
}

message LoginResponse {
  string session = 1;
}

message ListRequest {
  int32 offset = 1;
  int32 limit = 2;
}

message ListResponse {
  repeated string usernames = 1;
}

message CreateDocumentRequest {
  string name = 1;
  string doc = 2;
}

message CreateDocumentResponse {
  int64 id = 1;
  string token = 2;
}

message ListDocumentsRequest {
  int32 offset = 1;
  int32 limit = 2;
}

message ShortDocument {
  int64 id = 1;
  string name = 3;
}

message ListDocumentsResponse {
  repeated ShortDocument docs = 1;
}

message ExecuteRequest {
  string session = 1;
  string token = 2;
  int64 doc_id = 3;
}

message ExecuteResponse {
  string executed = 1;
}


message TestDocRequest {
  string session = 1;
  string content = 2;
}

message TestDocResponse {
  string executed = 1;
}

service Users {
  rpc Register(RegisterRequest) returns (RegisterResponse) {}
  rpc List(ListRequest) returns (ListResponse) {}
}

service Documents {
  rpc List(ListDocumentsRequest) returns (ListDocumentsResponse) {}
  rpc Create(CreateDocumentRequest) returns (CreateDocumentResponse) {}
  rpc Execute(ExecuteRequest) returns (ExecuteResponse) {}
  rpc TestDoc(TestDocRequest) returns (TestDocResponse) {}
}